README.rst
stxnext.grayscale
=================

Introduction
============

The ``stxnext.grayscale`` package is an add-on Product for Plone that gives the possibility of displaying
the web content of the site in the grayscale colors.

.. image:: https://github.com/stxnext/stxnext.grayscale/raw/master/docs/stxnext.grayscale.example.png

Colors transformation
=====================

The images are transformed on-the-fly using the PIL library.

The css and html documents are transformed with regular expression to substitute the colors
definitions with their grayscale equivalent values.

In Plone 4 (with plone.resource package installed) the transformed resources (images, css files)
are cached in the file system to avoid multiple transformation of the same resource.

Installation
============

Plone 3
-------

For sites based on Plone 3.x use `stxnext.grayscale v 1.0.0`_

Plone 4
-------

1. Add stxnext.grayscale to your plone.recipe.zope2instance section's eggs.
Define the resources parameter to point to the existing directory where the
transformed resources will be stored to avoid multiple transformation of
the same resource::

    [instance]
    recipe = plone.recipe.zope2instance
    ...
    resources = ${buildout:directory}/var/resources
    ...
    eggs =
        ...
        stxnext.grayscale
        
2. Install the Product via portal_quickinstaller

Configuration
=============

In the stxnext.grayscale tool's control panel (/@@grayscale_settings) choose the Plone skins
to be transformed to grayscale colors.

Author & Contact
================

 * Radosław Jankiewicz ``radoslaw.jankiewicz@stxnext.pl``
 * Sebastian Kalinowski ``sebastian.kalinowski@stxnext.pl``

.. image:: http://stxnext.pl/open-source/files/stx-next-logo

**STX Next** Sp. z o.o.

http://stxnext.pl

info@stxnext.pl

License
=======

This package is licensed under the Zope Public License.

.. _`stxnext.grayscale v 1.0.0`: https://pypi.python.org/pypi/stxnext.grayscale/1.0.0


docs/HISTORY.txt
Changelog
=========

2.0.0 (2013-11-20)
------------------

- Integration with Plone 4.3 [radekj]
- Executing the registered html transforms (diazo, tiles, ...) [radekj]
- Abandoned compatibility with Plone 3.x [radekj]


1.0.0 (2012-05-21)
------------------

- Initial release


src/stxnext/grayscale/tests/base.py

# -*- coding: utf-8 -*-
"""Base module for unittesting."""

import unittest

from plone.app.testing import FunctionalTesting
from plone.app.testing import IntegrationTesting
from plone.app.testing import PLONE_FIXTURE
from plone.app.testing import PloneSandboxLayer
from plone.app.testing import TEST_USER_NAME
from plone.app.testing import TEST_USER_ID
from plone.app.testing import login
from plone.app.testing import setRoles
from plone.app.testing import applyProfile


class GrayscaleLayer(PloneSandboxLayer):
    """Testing layer for stxnext.grayscale addon"""

    defaultBases = (PLONE_FIXTURE,)

    def setUpZope(self, app, configurationContext):
        """Set up Zope."""
        # Load ZCML
        import stxnext.grayscale

        self.loadZCML(package=stxnext.grayscale)

        from OFS.Application import install_package
        install_package(app, stxnext.grayscale, stxnext.grayscale.initialize)

    def setUpPloneSite(self, portal):
        """Set up Plone."""
        # import default profile
        applyProfile(portal, 'stxnext.grayscale:default')
        # Create test content
        # 1. Login as user with Manager privilages
        setRoles(portal, TEST_USER_ID, ['Manager'])
        login(portal, TEST_USER_NAME)

        # 2. Create and publish homepage object
        """
        portal.invokeFactory(id='homepage', type_name='GDACHomePage')
        homepage = portal.homepage
        homepage.setTitle('Homepage title')
        homepage.reindexObject()
        """

        import transaction
        transaction.commit()


    def tearDownZope(self, app):
        """Tear down Zope."""
        pass

FIXTURE = GrayscaleLayer()
INTEGRATION_TESTING = IntegrationTesting(
    bases=(FIXTURE,), name="GrayscaleLayer:Integration")
FUNCTIONAL_TESTING = FunctionalTesting(
    bases=(FIXTURE,), name="GrayscaleLayer:Functional")


class IntegrationTestCase(unittest.TestCase):
    """Base class for integration tests."""

    layer = INTEGRATION_TESTING

class FunctionalTestCase(unittest.TestCase):
    """Base class for functional tests."""

    layer = FUNCTIONAL_TESTING



src/stxnext/grayscale/tests/test_utils.py
import unittest

from stxnext.grayscale import utils


HEX_COLORS = """
border: 1px solid #fff;
border: 1pt solid #00FF00;
background: 123456;
background:#aaaaaa;
color:#1234;

style="color:#ff0"
"""
HEX_MATCH = sorted(['#fff', '#00FF00', '#aaaaaa', '#ff0'])

RGB_RGBA_COLORS = """
border: 1px solid rgb(10%, 20%, 10%);
border: 1pt solid rgba(11%,22%,33%, 0.1);
background: rgb(11,11);
background:rgba(44,55,66, 1);
color:rgb(44%,55%,66%);

style="color: rgb(1,2,3)"
"""
RGB_RGBA_MATCH = sorted(['rgb(10%, 20%, 10%)', 'rgba(11%,22%,33%, 0.1)',
    'rgb(11,11)', 'rgba(44,55,66, 1)', 'rgb(44%,55%,66%)', 'rgb(1,2,3)'])

NAMED_COLORS = """
border: 1px solid white; // will not be matched
border: 1pt solid black; // will not be matched
background-color: 123456;
color:       yellow;

style="color:red"
"""
NAMED_MATCH = sorted(['color:       yellow', 'color:red'])


class TestRegexpPatterns(unittest.TestCase):
    """
    Tests for regexp patterns
    """
    def test_match_hex_colors(self):
        """
        Tests COLOR_PATTERN for hex colors
        """
        colors = sorted(utils.COLOR_PATTERN.findall(HEX_COLORS))
        self.assertEqual(colors, HEX_MATCH)

    def test_match_rgb_rgba_colors(self):
        """
        Tests COLOR_PATTERN for RGB and RGBA colors
        """
        colors = sorted(utils.COLOR_PATTERN.findall(RGB_RGBA_COLORS))
        self.assertEqual(colors, RGB_RGBA_MATCH)

    def test_match_named_colors(self):
        """
        Tests COLOR_PATTERN for colors specified by name
        """
        colors = sorted(utils.COLOR_PATTERN.findall(NAMED_COLORS))
        self.assertEqual(colors, NAMED_MATCH)

    def test_dont_match_entities(self):
        """
        Tests if COLOR_PATTERN will not match HTML entities
        """
        entities = "&#60; &#160 &#182; &#8364; &#123456"
        self.assertFalse(utils.COLOR_PATTERN.findall(entities))


class TestAddParamToUrl(unittest.TestCase):
    """
    Test for stxnext.grayscale.utils.add_param_to_url function
    """

    def test_add_param_to_domain_wo_params(self):
        url_domain_only = 'http://example.com'
        self.assertEqual(utils.add_param_to_url(url_domain_only),
                'http://example.com?grayscale=1')

    def test_add_param_to_path_wo_params(self):
        url_with_path = 'http://example.com/abc/efg/'
        self.assertEqual(utils.add_param_to_url(url_with_path),
                'http://example.com/abc/efg/?grayscale=1')

    def test_add_param_to_file_wo_params(self):
        url_with_file = 'http://example.com/file.html'
        self.assertEqual(utils.add_param_to_url(url_with_file),
                'http://example.com/file.html?grayscale=1')

    def test_add_param_to_domain_w_params(self):
        url_domain_only = 'http://example.com?param1=2'
        self.assertEqual(utils.add_param_to_url(url_domain_only),
                'http://example.com?param1=2&grayscale=1')

    def test_add_param_to_path_w_params(self):
        url_with_path = 'http://example.com/abc/efg/?param1=2'
        self.assertEqual(utils.add_param_to_url(url_with_path),
                'http://example.com/abc/efg/?param1=2&grayscale=1')

    def test_add_param_to_file_w_params(self):
        url_with_file = 'http://example.com/file.html?param1=2'
        self.assertEqual(utils.add_param_to_url(url_with_file),
                'http://example.com/file.html?param1=2&grayscale=1')


class TestTransformImgSrc(unittest.TestCase):
    """
    Tests for stxnext.grayscale.utils.transform_img_src
    """
    def test_simple_img_tag(self):
        tag = """<img src="logo.png" />"""
        self.assertEqual(utils.transform_img_src(tag),
                """<img src="logo.png?grayscale=1" />""")

    def test_simple_img_tag_w_parmas(self):
        tag = """<img src="logo.png?param1=2" />"""
        self.assertEqual(utils.transform_img_src(tag),
                """<img src="logo.png?param1=2&grayscale=1" />""")

    def test_multiline_img_tag(self):
        tag = """<img
src="logo.png" />"""
        self.assertEqual(utils.transform_img_src(tag), """<img
src="logo.png?grayscale=1" />""")


class TestTransformCssUrl(unittest.TestCase):
    """
    Tests for stxnext.grayscale.utils.transform_css_url
    """
    def test_simple_css_url(self):
        css = """background: url(background.png)"""
        self.assertEqual(utils.transform_css_url(css),
                """background: url(background.png?grayscale=1)""")

    def test_simple_css_url_w_param(self):
        css = """background: url(background.png?param1=2)"""
        self.assertEqual(utils.transform_css_url(css),
                """background: url(background.png?param1=2&grayscale=1)""")

    def test_simple_css_url_w_quotation(self):
        css = """background: url("background.png")"""
        self.assertEqual(utils.transform_css_url(css),
                """background: url("background.png?grayscale=1")""")

    def test_simple_css_w_apostrophe(self):
        css = """background: url('background.png')"""
        self.assertEqual(utils.transform_css_url(css),
                """background: url('background.png?grayscale=1')""")


class TestTransformValue(unittest.TestCase):
    """
    Tests for stxnext.grayscale.utils.transform_value
    """
    def test_transform_named_color(self):
        """
        Test for standard named color
        """
        color = 'color: blue;'
        transformed = utils.transform_value(color)
        self.assertEqual(transformed, 'color: #1d1d1d')

    def test_transform_named_color_w_whitespaces(self):
        """
        Tests if named color with whitespaces will be transformed correctly
        """
        color = 'color:               gray        ;'
        transformed = utils.transform_value(color)
        self.assertEqual(transformed, 'color: #808080')

    def test_transform_other_named_color(self):
        """
        Tests if named color not included in utils.KWRD_MAP will be left unchanged
        """
        color = 'color: brown;'
        transformed = utils.transform_value(color)
        self.assertEqual(transformed, color)

    def test_transform_hex_6(self):
        """
        Tests if normal hex (6 chars) will be transformed correctly
        """
        color = '#ff0000'
        transformed = utils.transform_value(color)
        self.assertEqual(transformed, '#4c4c4c')

    def test_transform_hex_3(self):
        """
        Tests if normal hex (3 chars) will be transformed correctly
        """
        color = '#f00'
        transformed = utils.transform_value(color)
        self.assertEqual(transformed, '#4c4c4c')

    def test_transform_incorrect_len_hex(self):
        """
        Tests if hex with incorrect length will not be transformed
        """
        color = '#ff00'
        self.assertRaises(ValueError, utils.transform_value, color)

    def test_transform_incorrect_val_hex(self):
        """
        Tests if hex with incorrect values will not be transformed
        """
        color = '#qw0000'
        self.assertRaises(ValueError, utils.transform_value, color)

    def test_transform_rgb(self):
        """
        Tests rgb() color transformation
        """
        color = 'rgb(255, 0, 0)'
        transformed = utils.transform_value(color)
        self.assertEqual(transformed, '#4c4c4c')

    def test_transform_rgba(self):
        """
        Tests rgba() color transformation
        """
        color = 'rgba(255, 0, 0, 1)'
        transformed = utils.transform_value(color)
        self.assertEqual(transformed, 'rgba(76, 76, 76, 1.0)')


class TestGrayscaleTupe(unittest.TestCase):
    """
    Tests for stxnext.grayscale.utils.grayscale_tuple
    """

    def test_empty_tuple(self):
        self.assertRaises(ValueError, utils.grayscale_tuple, [])

    def test_icorrect_len_tuple(self):
        self.assertRaises(ValueError, utils.grayscale_tuple, [1,2])

    def test_incorrect_val_tuple(self):
        self.assertRaises(ValueError, utils.grayscale_tuple, ['a', 'b', 'c'])

    def test_not_tuple_or_list(self):
        self.assertRaises(ValueError, utils.grayscale_tuple, 'string')

    def test_correct_input(self):
        color = (255, 0, 0)
        self.assertEqual(utils.grayscale_tuple(color), (76, 76, 76))


class TestHexToTuple(unittest.TestCase):
    """
    Tests for stxnext.grayscale.utils.hex_to_tuple
    """

    def test_empty_text(self):
        self.assertRaises(ValueError, utils.hex_to_tuple, '')

    def test_incorrect_len_text(self):
        self.assertRaises(ValueError, utils.hex_to_tuple, '#a')

    def test_incorrect_val_text(self):
        self.assertRaises(ValueError, utils.hex_to_tuple, '#qwe')

    def test_correct_input_6(self):
        text = '#ff0000'
        self.assertEqual(utils.hex_to_tuple(text), (255, 0, 0))

    def test_correct_input_3(self):
        text = '#f00'
        self.assertEqual(utils.hex_to_tuple(text), (255, 0, 0))

    def test_input_w_whitespace(self):
        text = '             #ff0000             '
        self.assertEqual(utils.hex_to_tuple(text), (255, 0, 0))

    def test_input_wo_hash(self):
        text = 'ff0000'
        self.assertEqual(utils.hex_to_tuple(text), (255, 0, 0))


class TestRgbToTuple(unittest.TestCase):
    """
    Tests for stxnext.grayscale.utils.rgb_to_tuple
    """

    def test_correct_numbers_input(self):
        text = 'rgb(1,2,3)'
        self.assertEqual(utils.rgb_to_tuple(text), (1, 2, 3))

    def test_correct_percentages_input(self):
        text = 'rgb(100%, 0, 0)'
        self.assertEqual(utils.rgb_to_tuple(text), (255, 0, 0))

    def test_not_rgb_input(self):
        text = 'some string'
        self.assertEqual(utils.rgb_to_tuple(text), ())


class TestRgbaGrayscake(unittest.TestCase):
    """
    Tests for stxnext.grayscale.utils.rgba_grayscale
    """

    def test_incorrect_len_input(self):
        """
        Tests for situtation when (for example) rgb() will be passed
        """
        text = 'rgb(1, 2, 3)'
        self.assertRaises(ValueError, utils.rgba_grayscale, text)

    def test_input_w_float_alpha(self):
        text = 'rgba(255, 0, 0, 0.1)'
        self.assertEqual(utils.rgba_grayscale(text), 'rgba(76, 76, 76, 0.1)')

    def test_input_w_int_alpha(self):
        text = 'rgba(255, 0, 0, 1)'
        self.assertEqual(utils.rgba_grayscale(text), 'rgba(76, 76, 76, 1.0)')


class TestAddBodyclass(unittest.TestCase):
    """
    Tests for stxnext.grayscale.utils.add_bodyclass
    """

    def test_body_no_attrs(self):
        text = """<body>"""
        self.assertEqual(utils.add_bodyclass(text),
                """<body class="gray-style">""")

    def test_body_w_attrs(self):
        text = """<body dir="ltr">"""
        self.assertEqual(utils.add_bodyclass(text),
                """<body dir="ltr" class="gray-style">""")

    def test_body_w_class_no_other_attrs(self):
        text = """<body class='homepage'>"""
        self.assertEqual(utils.add_bodyclass(text),
                """<body class='gray-style homepage'>""")

    def test_body_w_class_many_no_other_attrs(self):
        text = """<body class="homepage template-document_view">"""
        self.assertEqual(utils.add_bodyclass(text),
                """<body class="gray-style homepage template-document_view">""")

    def test_body_w_class_w_other_attrs(self):
        text = """<body class='homepage template-document_view' dir='ltr'>"""
        self.assertEqual(utils.add_bodyclass(text),
                """<body class='gray-style homepage template-document_view' dir='ltr'>""")


def test_suite():
    return unittest.defaultTestLoader.loadTestsFromName(__name__)


src/stxnext/grayscale/events.py
# -*- coding:utf-8 -*-
from Acquisition import aq_base
from OFS.Image import File
from Products.CMFCore.FSFile import FSFile
from Products.CMFCore.FSImage import FSImage
from Products.ATContentTypes.interface.image import IATImage
from Products.ATContentTypes.content.file import ATFile
from plone.app.linkintegrity.interfaces import IOFSImage
from zope.app.component.hooks import getSite
from zope.publisher.interfaces.browser import IBrowserView
from zope.component import queryUtility
from zope.component.interfaces import ComponentLookupError
from Products.CMFCore.utils import getToolByName
from Products.CMFCore.interfaces import ISiteRoot
from Products.CMFCore.FSPageTemplate import FSPageTemplate
from zExceptions import NotFound

from stxnext.grayscale import log
import utils

# compatibility with Plone 3
try:
    from plone.resource.file import FilesystemFile
    from plone.resource.interfaces import IResourceDirectory
    PLONE_RESOURCE_INSTALLED = True
except ImportError:
    PLONE_RESOURCE_INSTALLED = False
try:
    from zope.browserresource.file import File as browserresourcefile
    BROWSER_RESOURCE_INSTALLED = True
except ImportError:
    BROWSER_RESOURCE_INSTALLED = False



def transformation_event(event):
    """
    Checks if current default theme is 
    meant to be transformed to grayscale
    """
    request = event.request
    portal = getSite()
    if not portal:
        return
    properties_tool = getToolByName(portal, 'portal_properties')
    transformed_themes = properties_tool.site_properties.getProperty('transformed_themes')
    if not transformed_themes:
        return
    skins_tool = getToolByName(portal, 'portal_skins')
    request_varname = skins_tool.request_varname
    if request.get(request_varname):
        selected_skin = request.get(request_varname)
    else:
        selected_skin = skins_tool.getDefaultSkin()
    if selected_skin in transformed_themes:
        GrayscaleTransformations(event)

def GrayscaleTransformations(event):
    """
    Changing the response body by converting
    the images to grayscale and changing the
    css colors definitions to colors from gray
    shades palette
    """
    request = event.request
    response = request.response
    context = request.get('PUBLISHED')
    resp_body = response.getBody()
    content_type = ''
    path = ''

    resp_body = getattr(context, 'GET', lambda: '')()
    if hasattr(context, 'im_self'):
        context = context.im_self
    if hasattr(context, 'context'):
        context = context.context
    
    if PLONE_RESOURCE_INSTALLED and isinstance (context, FilesystemFile):
        resp_body = context().read()
        path = context.path
        content_type = context.getContentType().split(';')[0]
    
    filename = getattr(context, 'getId', lambda: False)()
    if not filename:
        try:
            filename = context.__name__
        except AttributeError:
            return
        
    images_content_types = ['image/png', 'image/jpg', 'image/jpeg', 'image/gif']
    
    browser_resource_image = False
    if BROWSER_RESOURCE_INSTALLED:
        if isinstance (context, browserresourcefile) and \
           context.content_type.split(';')[0] in images_content_types:
            browser_resource_image = True
        
    if isinstance (context, FSImage) or \
       IOFSImage.providedBy(context) or \
       IATImage.providedBy(context) or \
       content_type in images_content_types or \
       browser_resource_image:
        
        if not path:
            try:
                path = '/'.join(context.getPhysicalPath())
            except AttributeError:
                path = filename
                        
        try:
            if PLONE_RESOURCE_INSTALLED:
                resp_body = utils.get_resource(request, response, filename)
            else:
                raise NotFound
        except NotFound:
            image_body = resp_body
            if not image_body:
                if hasattr(context, 'data'):
                    image_body = context.data
                elif isinstance (context, FSImage):
                    image_body = context._readFile(True)
                    
            if image_body:
                resp_body = utils.image_to_grayscale(image_body, path)
            else:
                log.debug('Image doesn\'t contain any data: %s' % (path))
                
            if PLONE_RESOURCE_INSTALLED:
                if queryUtility(IResourceDirectory, name=u''):
                    utils.store_resource(filename, resp_body)
        
    elif IBrowserView.providedBy(request.get('PUBLISHED')) or \
         content_type in ['text/html', 'text/css'] or \
         isinstance(context, (File, FSFile, ATFile, FSPageTemplate)) and \
         context.content_type.split(';')[0] in ['text/html', 'text/css']:
        
        if hasattr(aq_base(context), 'data'):
            resp_body = context.data
            if hasattr(resp_body, 'data'):
                resp_body = resp_body.data
                
        if isinstance(context, FSFile):
            resp_body = context._readFile(0)
        
        if content_type == 'text/css' or \
           isinstance(context, (File, FSFile, ATFile, FSPageTemplate)) and \
           context.content_type.split(';')[0] == 'text/css':
            
            try:
                if PLONE_RESOURCE_INSTALLED:
                    resp_body = utils.get_resource(request, response, filename)
                else:
                    raise NotFound
            except NotFound:                
                resp_body = utils.transform_style_properties(resp_body)
                resp_body = utils.transform_css_url(resp_body)
                if PLONE_RESOURCE_INSTALLED:
                    if queryUtility(IResourceDirectory, name=u''):
                        utils.store_resource(filename, resp_body)
                
        else:
            if not resp_body:
                resp_body = utils.render_object_html(context, request)
            resp_body = utils.add_bodyclass(resp_body)
            resp_body = utils.transform_style_properties(resp_body)
            resp_body = utils.transform_css_url(resp_body)
            resp_body = utils.transform_img_src(resp_body)
            
    response.setBody(resp_body)
    


src/stxnext/grayscale/__init__.py
# -*- coding: utf-8 -*-
import os
import logging
from zope.i18nmessageid import MessageFactory
from zope.component import queryUtility
try:
    from plone.resource.interfaces import IResourceDirectory
    PLONE_RESOURCE_INSTALLED = True
except ImportError:
    PLONE_RESOURCE_INSTALLED = False
    
from stxnext.grayscale.config import TYPE, THEME

log = logging.getLogger('stxnext.grayscale')
StxnextGrayscaleMessageFactory = MessageFactory('stxnext.grayscale')

def initialize(context):
    """
    Initializer called when used as a Zope 2 product.
    """
    if PLONE_RESOURCE_INSTALLED:
        resource_dir = queryUtility(IResourceDirectory, name=u'')
        if resource_dir:
            if not TYPE in resource_dir.listDirectory():
                path = os.path.join(resource_dir.directory, TYPE)
                os.makedirs(os.path.normpath(path))
                log.info("Created directory for stroring grayscaled resources: %s" % path)
            grayscale_dir = resource_dir[TYPE]
            if not THEME in grayscale_dir.listDirectory():
                path = os.path.join(grayscale_dir.directory, THEME)
                os.makedirs(os.path.normpath(path))
                log.info("Created directory for stroring grayscaled resources: %s" % path)

src/stxnext/grayscale/browser/preferences/grayscale_settings.py
# -*- coding:utf-8 -*-
import os

from zope.app.component.hooks import getSite
from zope.component import adapts
from zope.event import notify
from zope.interface import Interface, implements
from zope.formlib.form import action, applyChanges, FormFields
from zope.schema import Bool, Choice, Tuple

from Products.CMFDefault.formlib.schema import SchemaAdapterBase, ProxyFieldProperty
from Products.CMFPlone.interfaces import IPloneSiteRoot
from Products.statusmessages.interfaces import IStatusMessage
from plone.app.controlpanel.events import ConfigurationChangedEvent
from plone.app.controlpanel.form import ControlPanelForm
from plone.app.controlpanel.site import SiteControlPanel, SiteControlPanelAdapter
from plone.protect import CheckAuthenticator

from stxnext.grayscale import log
from stxnext.grayscale import StxnextGrayscaleMessageFactory as _
from stxnext.grayscale.config import TYPE, THEME


class IGrayscaleSettingsSchema(Interface):
    """
    Interface for additional control panel settings
    """

    transformed_themes = Tuple(
        title = _(u'Skins transformed to grayscale'),
        description = _(u'Select skin to be transformed to grayscale'),
        required = False,
        value_type = Choice(required = False,
            vocabulary = 'plone.app.vocabularies.Skins')
       )
    
    delete_cached_resources = Bool(
        title = _(u'Delete cached resources'),
        description=_(u'Select the checkbox to delete the gray-scale ' \
                      u'converted resources from the filesystem cache.'),
       )

class GrayscaleControlPanelAdapter(SiteControlPanelAdapter):
    """
    Storage for grayscale site control panel settings
    """

    adapts(IPloneSiteRoot)
    implements(IGrayscaleSettingsSchema)

    transformed_themes = ProxyFieldProperty(IGrayscaleSettingsSchema['transformed_themes'])
    delete_cached_resources = False

class GrayscaleSettingsForm(ControlPanelForm):
    """
    Grayscale transformation skins settings form
    """

    form_fields = FormFields(IGrayscaleSettingsSchema)

    label = _('Grayscale settings')
    description = ''
    form_name = _('Grayscale skin settings')
    
    def setUpWidgets(self, ignore_request=False):
        """
        Disabling the widget for removing filestorage cache in case of
        lack of grayscale resources dir registered
        """
        super(GrayscaleSettingsForm, self).setUpWidgets(ignore_request=ignore_request)
        dcr_widget = self.widgets['delete_cached_resources']
        site = getSite()
        path = str('/++%s++%s' % (TYPE, THEME))
        try:
            location = site.restrictedTraverse(path)
        except AttributeError:
            dcr_widget.disabled = True
    
    @action(_(u'label_save', default=u'Save'), name=u'save')
    def handle_edit_action(self, action, data):
        CheckAuthenticator(self.request)
        if applyChanges(self.context, self.form_fields, data, self.adapters):
            notify(ConfigurationChangedEvent(self, data))
            return self._on_save(data)

    def _on_save(self, data):
        """
        Deleting the gray-scale converted resources
        from the filesystem cache
        """
        messages = IStatusMessage(self.request)
        if data.get('delete_cached_resources'):
            site = getSite()
            try:
                path = str('/++%s++%s' % (TYPE, THEME))
                location = site.restrictedTraverse(path).directory
            except AttributeError, e:
                log.info('Unable to remove files for: %s due to: AttributeError: %s' % (path, e))
                message = _(u"Couldn't remove files from %s" % path)
                messages.addStatusMessage(message, type='error')
                return

            try:
                log.info('Removing cached gray-scale files from: %s' % location)
                for filename in os.listdir(location):
                    file_path = os.path.join(location, filename)
                    if os.path.isfile(file_path):
                        os.unlink(file_path)
            except OSError, e:
                log.info('Unable to remove files from: %s due to error: %s' % (location, e))
                message = _(u"Couldn't remove files from %s" % location)
                messages.addStatusMessage(message, type='error')
            else:
                message = _(u'Files from path %s have been successfully removed.' % location)
                messages.addStatusMessage(message, type='info')


src/stxnext/grayscale/utils.py
# -*- coding:utf-8 -*-
import os
import re
import urlparse
from PIL import Image
from StringIO import StringIO

from zExceptions import NotFound
from zope.app.component.hooks import getSite
from zope.component import queryMultiAdapter
from zope.contentprovider.interfaces import ContentProviderLookupError
from Products.CMFCore.FSPageTemplate import FSPageTemplate
from Products.Five import BrowserView
from Products.PythonScripts.PythonScript import PythonScript

from stxnext.grayscale import log
from stxnext.grayscale.config import TYPE, THEME

COLOR_PATTERN = re.compile(r"(?<=[:\s])#[a-fA-F0-9]{6}|(?<=[:\s])#[a-fA-F0-9]{3}(?=[\s;\"\'])|rgba?\(.+?\)|color:\s*(?!rgb)[a-z]+", re.M | re.I)
IMG_SRC_PATTERN = re.compile(r"<\s*img\s+[^>]*src\s*=\s*[\"']?([^\"' >]+)[\"']", re.I)
CSS_URL_PATTERN = re.compile(r"""url\(["']?([^\)'"]+)['"]?\)""", re.I)

# most popular css color names and their grayscale versions in css notation.
KWRD_MAP = {
  'blue'   : '#1d1d1d',
  'gray'   : '#808080',
  'green'  : '#969696',
  'orange' : '#8c8c8c',
  'purple' : '#555555',
  'red'    : '#4c4c4c',
  'silver' : '#c0c0c0',
  'yellow' : '#aaaaaa',
  }


def add_param_to_url(url):
    """
    Adds "&grayscale=1" param to the url
    """
    url = list(urlparse.urlsplit(url))
    if url[3]:
        url[3] += '&grayscale=1'
    else:
        url[3] = 'grayscale=1'
    return urlparse.urlunsplit(url)

def transform_img_src(text):
    """
    Adds "grayscale=1" param to the ULR in src attribute in <img> tags to avoid
    caching images
    """
    urls = IMG_SRC_PATTERN.findall(text)
    for url in urls:
        new_url = add_param_to_url(url)
        text = text.replace(url, new_url)
    return text

def transform_css_url(text):
    """
    Adds "grayscale=1" param to the URL in url() directive in CSS to avoid
    caching images
    """
    urls = CSS_URL_PATTERN.findall(text)
    for url in urls:
        new_url = add_param_to_url(url)
        text = text.replace(url, new_url)
    return text

def transform_style_properties(text):
    """
    Converts the occurances of the color definitions
    to gray scale equivalents
    """
    matches = COLOR_PATTERN.findall(text)
    for match in set(matches):
        converted_match = transform_value(match)
        text = text.replace(match, converted_match)
    return text

def transform_value(color_str):
    """
    Transforms a single color value given
    in any string format to a relevant grayscale
    representation.
    """
    color_val = color_str.split(':')[-1].lower().strip(' ;')
    if color_val in KWRD_MAP:
        return 'color: %s' % KWRD_MAP[color_val]
    elif not color_str.startswith('color'):
        if color_val.startswith('#'):
            tup_val = hex_to_tuple(color_val)
        elif color_val.startswith('rgba'):
            return rgba_grayscale(color_val)
        elif color_val.startswith('rgb'):
            tup_val = rgb_to_tuple(color_val)
        else:
            return color_str
        return tuple_to_hex(grayscale_tuple(tup_val))
    else:
        return color_str

def grayscale_tuple(tup):
    """
    Takes a 3-tuple and returns a 3-tuple,
    with the valu of the input color converted
    to grayscale
    """
    try:
        gray = 0.299 * tup[0] + 0.587 * tup[1] + 0.114 * tup[2]
        gray = int(round(gray))
    except (IndexError, TypeError):
        raise ValueError, "input %s is not in (r, g, b) format" % str(tup)
    return (gray, gray, gray)

def hex_to_tuple(text):
    """
    Transform #rrggbb notation of color 
    to 3-tuple (r, g, b)
    """
    colorstring = text.strip()
    # remove trailing "#"
    colorstring = colorstring.lstrip('#')
    if len(colorstring) == 3:
        r, g, b = (colorstring[0] * 2, colorstring[1] * 2, colorstring[2] * 2)
    elif len(colorstring) == 6:
        r, g, b = colorstring[:2], colorstring[2:4], colorstring[4:]
    else:
        raise ValueError, "input #%s is not in #RRGGBB format" % colorstring
    return tuple([int(x, 16) for x in (r, g, b)])

def rgb_to_tuple(text):
    """
    Transforms rgb(x,x,x) both in decimal and %
    notation of color to 3-tuple (r, g, b)
    """
    numbers = re.findall(r'\d+', text)
    if '%' in text:
        tup = tuple([int((int(x)/100.0) * 255) for x in numbers])
    else:
        tup = tuple([int(x) for x in numbers])
    return tup

def rgba_grayscale(text):
    """
    Replaces x, y, z colors in rgba(x,y,z,a) with coresponding grayscale values
    """
    tup_val = rgb_to_tuple(text)
    rgb_grayscale_tup = grayscale_tuple(tup_val[:3])
    if not tup_val[3:5]:
        raise ValueError, "input %s is not in rgba(x,y,z,a) format" % text
    alpha = float('.'.join([str(n) for n in tup_val[3:5]]))
    rgba_grayscale_tup = rgb_grayscale_tup + (alpha, )
    return "rgba(%d, %d, %d, %.1f)" % rgba_grayscale_tup

def tuple_to_hex(tup):
    """
    Transforms 3-tuple with colors to 
    #RRGGBB html notation.
    """
    return '#%02x%02x%02x' % tup

def image_to_grayscale(image_data, path):
    """
    Converts image data to grayscale using PIL library 
    """
    try:
        imagestring = StringIO(image_data)
        image_file = Image.open(imagestring)
        converted = image_file.convert('RGBA')
        converted = converted.convert('LA')
        res = StringIO()
        converted.save(res, format='PNG')
        return res.getvalue()
    except IOError, err:
        log.error('Error while transforming image: %s : %s' % (path, err))
        return image_data

def add_bodyclass(html_text):
    """
    Applies the 'gray-scale' css class on body tag 
    """
    bodymatches = re.findall(r"<body.*>", html_text, re.I)
    for bodytag in bodymatches:
        if 'class' in bodytag:
            for css_classs in re.findall(r"""class\s*=\s*[\'\"]([-\w ]*)[^>]*>""", bodytag, re.I):
                modified_bodytag = bodytag.replace(css_classs,
                    ' '.join(('gray-style', css_classs)))
        else:
            modified_bodytag = bodytag.replace('>', ' class="gray-style">')
        html_text = html_text.replace(bodytag, modified_bodytag)
    return html_text

def get_resource(request, response, filename):
    """
    Returns the data of the resource cached
    in the file system
    """
    site = getSite()
    cached_file = site.restrictedTraverse(str('/++%s++%s/%s' % (TYPE, THEME, filename)))
    return cached_file(REQUEST=request, RESPONSE=response).read()

def store_resource(filename, data):
    """
    Stores the data of the resource in the
    file system
    """
    site = getSite()
    location = site.restrictedTraverse(str('/++%s++%s' % (TYPE, THEME))).directory
    fs_file = open(os.path.join(location, filename), "w")
    try:
        fs_file.write(data)
    finally:
        fs_file.close()

def render_object_html(obj, request):
    """
    Returns rendered html for given content object
    """
    published = request.get('PUBLISHED')
    if isinstance(published, (BrowserView, FSPageTemplate, PythonScript)):
        try:
            return published() or ''
        except NotFound:
            log.error("Resource '%s' not found" % repr(obj))
            return ''
    
    def_page_id = obj.getDefaultPage()
    if def_page_id:
        def_page = obj[def_page_id]
        return render_object_html(def_page, request)

    view_name = obj.getLayout()
    view = queryMultiAdapter((obj, request), name=view_name)
    if view:
        try:
            return view.context() or ''
        except ContentProviderLookupError:
            pass

    view = obj.restrictedTraverse(view_name, None)
    if view:
        try:
            return view.context() or ''
        except AttributeError:
            return view() or ''

    try:
        return obj()
    except AttributeError:
        pass
    return ''
    
